Allow routes from routes prop

alexanderchr opened the pull request:
As discussed in #252.

alexanderchr commented:
Might actually have to do this for all child routes as well, will test.

gaearon commented on src/patch.dev.js:
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
+  if (isReactRouterish(type) && props) {
+    // A router's routes may be given as children or through
+    // the `routes` prop.
+    var topRoutes = [].concat(props.routes, props.children).filter(Boolean);
```
Let’s move this initial loop into `forceUpdateRouteAndChildRouteComponents` as well.

alexanderchr replied:
Sure.

gaearon commented on src/patch.dev.js:
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
+  // elements and in `childRoutes` when using plain routes
+  var childRoutes = [].concat(route.childRoutes, route.props && route.props.children).filter(Boolean);
+
+  for (var _len = childRoutes.length, _key = 0; _key < _len; _key++) {
```
Let’s just use `for (var i = 0; i < childRoutes.length; i++)`?

gaearon commented on src/patch.dev.js:
```diff
@@ -77,26 +77,42 @@ if (React.createElement.isPatchedByReactHotLoader) {
+    .filter(Boolean);
+
+  for (var i = 0; i < childRoutes.length; i++) {
+    var childRoute = childRoutes[_key];
```
Still using `_key` here

alexanderchr replied:
Yeah, nice catch. It seems like it worked because babel were transpiling it to _key anyways.

gaearon commented on src/patch.dev.js:
```diff
@@ -77,26 +77,42 @@ if (React.createElement.isPatchedByReactHotLoader) {
+
+  // Child routes will be in `props.children` when defining routes using react
+  // elements and in `routes` or `childRoutes` when using plain routes
+  var childRoutes = []
```
Minor nit: let’s use `const` where possible, `let` otherwise.

alexanderchr replied:
Sure.

gaearon commented on src/patch.dev.js:
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
+function forceUpdateComponentOfRouteAndChildRoutes(route) {
+  // TODO: check whether it is possible to also handle the `getComponent` case here
+  const components = []
+    .concat(route.component, route.components && Object.values(route.components))
```
`Object.values` may not work in older browsers, we shoot for wide compatibility.

alexanderchr replied:
Right, changed it a bit.

gaearon commented on src/patch.dev.js:
```diff
@@ -77,26 +77,43 @@ if (React.createElement.isPatchedByReactHotLoader) {
+    .concat(route.component, route.components && Object.values(route.components))
+    .filter(c => typeof(c) === 'function');
+
+  for(const component of components) {
```
Nit: space after `for`.

gaearon commented:
Thanks, this looks good. I have a few more style nits but I’m happy to fix those myself. I need to add a linter anyway.

Would you be open to considering adding some tests for this? I have a basic Mocha setup for the Babel plugin, and it seems like this is another example of something we want to be tested.

alexanderchr commented on src/hackRouter.js:
```diff
@@ -0,0 +1,48 @@
+  }
+  return values;
+}
+
```
Might want to use a real polyfill instead

gaearon replied:
Maybe we’d just use `Object.keys(obj).map(key => obj[key])`?

alexanderchr replied:
Yeah, that's it. Thanks.

alexanderchr commented:
Sure, added a few tests.

Also had some midnight inspiration and rewrote all of it. I included these changes in the last commit, so feel free to discard them if you think the old way was better. Otherwise you can just squash them into the first refactor commit. Tests should hopefully be passing either way.

gaearon commented:
This looks great! `hackRouter` sounds a little intimidating, let’s rename it to `addReactRouterFix()` or something similar?

alexanderchr commented:
Sure!

gaearon commented on src/fixupReactRouter.js:
```diff
@@ -0,0 +1,52 @@
+  // https://github.com/reactjs/react-router/issues/2182
+  const routes = props.routes || ensureArray(props.children);
+  const components = componentsToForceUpdate(routes);
+  for (const component of components) {
```
[`for of` requires `Symbol`s](http://babeljs.io/repl/#?evaluate=true&presets=es2015%2Creact&experimental=false&loose=false&spec=false&code=for%20(var%20x%20of%20y)%20%7B%0A%20%20%0A%7D), let’s use `components.forEach`

gaearon commented on src/fixupReactRouter.js:
```diff
@@ -0,0 +1,52 @@
+  );
+}
+
+function componentsToForceUpdate(routes) {
```
Let’s rename to `extractComponents`. We’re not exactly force updating them anyway.

gaearon commented:
Out in `3.0.0-alpha.13`, thank you!

alexanderchr commented:
Cool. Thanks for reviewing!
